cmd/loom/loom.go

```diff
@@ -644,14 +647,38 @@ func loadAppStore(cfg *config.Config, logger *loom.Logger, targetVersion int64)
 		if err != nil {
 			return nil, err
 		}
+		appStoreEvmRoot := iavlStore.Get(rootKey)
 		evmStore, err := loadEvmStore(cfg, iavlStore.Version())
 		if err != nil {
 			return nil, err
 		}
+		evmStoreEvmRoot, evmVersion := evmStore.GetLastSavedRoot(iavlStore.Version())
+		if !bytes.Equal(appStoreEvmRoot, evmStoreEvmRoot) {
+			// downgrade iavlStore version
+			log.Info(
+				"EVM root mismatch, resyncing roots...",
+				"evm.db-ver", evmVersion, "app.db-ver", iavlStore.Version(),
+				"evm.db-root", hex.EncodeToString(evmStoreEvmRoot), "app.db-evm-root", hex.EncodeToString(appStoreEvmRoot),
+			)
+			lastestVersion, _ := store.LastestIAVLStoreVersion(db, evmVersion)
```

Comment:
Normally, function `MutableTree.LoadVersion` itself will return the latest version available that's less than the target version if the target version is not available along with the error.
```golang
	if !(targetVersion == 0 || latestVersion == targetVersion) {
		return latestVersion, fmt.Errorf("wanted to load target %v but only found up to %v",
			targetVersion, latestVersion)
	}
```
for example : `latestVersion,errror = 10, "wanted to load target 25 but only found up to 10"` 

And my intension for these line of code 
```golang
if lastestVersion == 0 {
 lastestVersion = -1
 }
```
Is if the function `MutableTree.LoadVersion` find out that the latest version available is version 0, And we continue call `store.NewIAVLStore` function with target version = 0
it will automatically return `IAVLStore` with Newest version not the version 0 that we want.
This is what I do to solve the problem.
```golang
func NewIAVLStore(db dbm.DB, maxVersions, targetVersion, flushInterval int64) (*IAVLStore, error) {
	tree := iavl.NewMutableTree(db, 10000)
	if targetVersion != -1 {
		_, err := tree.LoadVersion(targetVersion)
		if err != nil {
			return nil, err
		}
	}
	// always keep at least 2 of the last versions
	if (maxVersions != 0) && (maxVersions < 2) {
		maxVersions = 2
	}

	return &IAVLStore{
		tree:          tree,
		maxVersions:   maxVersions,
		flushInterval: flushInterval,
	}, nil
}
```
I just skip `tree.LoadVersion` from loading latest version again.

Now I will try to move to `store/multi_writer_app_store.go` as you comment.

Reply:
> Normally, function `MutableTree.LoadVersion` itself will return the latest version available that's less than the target version if the target version is not available along with the error.

If it returns an error you can't rely on the return value, otherwise you're relying on internal implementation details that could change in the next version of the package.

Reply:
got it.